modyn/storage/src/internal/grpc/storage_service_impl.cpp

```diff
@@ -0,0 +1,596 @@
+    while (!sample_ids_futures_queue.empty()) {
+      T response;
+
+      SampleData sample_data = sample_ids_futures_queue.front().get();
+      sample_ids_futures_queue.pop();
+
+      for (size_t i = 0; i < sample_data.ids.size(); i++) {
+        response.add_keys(sample_data.ids[i]);
+        response.add_labels(sample_data.labels[i]);
+      }
+
+      writer->Write(response);
+    }
+  }
+}
+
+SampleData StorageServiceImpl::get_sample_subset(
+    int64_t file_id, int64_t start_index, int64_t end_index,
+    const storage::database::StorageDatabaseConnection& storage_database_connection) {
+  soci::session session = storage_database_connection.get_session();
+  const int64_t number_of_samples = end_index - start_index + 1;
+  std::vector<int64_t> sample_ids(number_of_samples + 1);
+  std::vector<int64_t> sample_labels(number_of_samples + 1);
+  session << "SELECT sample_id, label FROM samples WHERE file_id = :file_id AND sample_index >= :start_index AND "
+             "sample_index "
+             "<= :end_index",
+      soci::into(sample_ids), soci::into(sample_labels), soci::use(file_id), soci::use(start_index),
+      soci::use(end_index);
+  return {sample_ids, {}, sample_labels};
+}
+
+int64_t StorageServiceImpl::get_number_of_samples_in_file(int64_t file_id, soci::session& session) {
+  int64_t number_of_samples;
+  session << "SELECT number_of_samples FROM files WHERE file_id = :file_id", soci::into(number_of_samples),
+      soci::use(file_id);
+  return number_of_samples;
+}
+
+std::tuple<int64_t, int64_t> StorageServiceImpl::get_partition_for_worker(int64_t worker_id, int64_t total_workers,
+                                                                          int64_t total_num_elements) {
+  if (worker_id < 0 || worker_id >= total_workers) {
+    FAIL("Worker id must be between 0 and total_workers - 1.");
+  }
+
+  const int64_t subset_size = total_num_elements / total_workers;
+  int64_t worker_subset_size = subset_size;
+
+  const int64_t threshold = total_num_elements % total_workers;
+  if (threshold > 0) {
+    if (worker_id < threshold) {
+      worker_subset_size += 1;
+      const int64_t start_index = worker_id * (subset_size + 1);
+      return {start_index, worker_subset_size};
+    }
+    const int64_t start_index = threshold * (subset_size + 1) + (worker_id - threshold) * subset_size;
+    return {start_index, worker_subset_size};
+  }
+  const int64_t start_index = worker_id * subset_size;
+  return {start_index, worker_subset_size};
+}
+
+int64_t StorageServiceImpl::get_dataset_id(const std::string& dataset_name, soci::session& session) {
+  int64_t dataset_id = -1;  // NOLINT misc-const-correctness
+  session << "SELECT dataset_id FROM datasets WHERE name = :name", soci::into(dataset_id), soci::use(dataset_name);
+
+  return dataset_id;
+}
+
+std::vector<int64_t> StorageServiceImpl::get_file_ids(int64_t dataset_id, soci::session& session,
```

Comment:
Uff, this is _very_ hard to read. Before I try to review this, could you try to split the cases here into several different submethods? Because right now this is visual sensory overload :D